### PR TITLE
refactor: use `_MiB`/`_GiB` consistently for byte conversions

### PR DESCRIPTION
**Pull request by l0rinc**
### Problem
Byte-size conversions in the codebase currently show up in many equivalent formats (multiplication/division chains, shifts, hex/binary literals), which creates a maintenance burden and makes review error-prone - especially considering the architectural differences of `size_t`.
Inspired by https://github.com/bitcoin/bitcoin/pull/34305#discussion_r2734720002, it seemed appropriate to unify `Mebibyte` usage across the codebase and add `Gibibyte` support with 32/64 bit `size_t` validation.

### Fix
This PR refactors those call sites to use `""_MiB` (existing) and `""_GiB` (new), and adds the encountered value/pattern replacements to unit tests to make review straightforward, and to ensure the conversions remain valid.
The literals are overflow-checked when converting to `size_t`, and unit tests cover the 32-bit boundary cases.

Concretely, it replaces patterns such as:
* `1024*1024`, `1<<20`, `0x100000`, `1048576`, `/ 1024 / 1024`, `* (1.0 / 1024 / 1024)` → `1_MiB` or `double(1_MiB)`
* `1024*1024*1024`, `1<<30`, `0x40000000`, `1024_MiB`, `>> 30` → `1_GiB`

(added unit tests for each replacement category to ease review)

Additionally, declarations whose initializer reads a `_MiB`/`_GiB` literal are switched to braced initialization so a future oversized value is rejected at compile time through the narrowing check instead of silently truncating.

### Note
In the few places where arithmetic involves signed values, the result is identical to the previous code assuming those quantities never become negative.

### CODE REVIEW

File: src/util/byte_units.h [hunk 1]
```diff
@@ -7,16 +7,32 @@
...
 
+#include <limits>
+#include <string>
 #include <stdexcept>
 
-//! Overflow-safe conversion of MiB to bytes.
-constexpr size_t operator""_MiB(unsigned long long mebibytes)
+namespace util::detail {
+template<unsigned SHIFT>
+constexpr size_t ByteUnitsToBytes(unsigned long long units, const std::string& exception_msg)
```
Comment (hodlinator) [inline src/util/byte_units.h]: Should use `string_view` or `const char*` to avoid heap allocations for the happy path.
```suggestion
constexpr size_t ByteUnitsToBytes(unsigned long long units, std::string_view exception_msg)
```

(Edit: I realize this is only run in a constexpr context so it's probably negligible in practice, but would still be nice to fix).

File: src/qt/optionsmodel.h [hunk 2]
```diff
@@ -26,12 +27,12 @@ static constexpr uint16_t DEFAULT_GUI_PROXY_PORT = 9050;
 /**
  * Convert configured prune target MiB to displayed GB. Round up to avoid underestimating max disk usage.
  */
-static inline int PruneMiBtoGB(int64_t mib) { return (mib * 1024 * 1024 + GB_BYTES - 1) / GB_BYTES; }
+static inline int PruneMiBtoGB(int64_t mib) { return (mib * 1_MiB + GB_BYTES - 1) / GB_BYTES; }
```
Comment (hodlinator) [inline src/qt/optionsmodel.h:30]: I wonder if multiplying with `size_t` instead of unspecified integer expression risks flipping negative function inputs into positive output values, but it's probably only used with positive inputs regardless.

File: src/util/byte_units.h [hunk 3]
```diff
@@ -7,16 +7,32 @@
...
 #include <util/overflow.h>
 
+#include <limits>
+#include <string>
 #include <stdexcept>
 
-//! Overflow-safe conversion of MiB to bytes.
-constexpr size_t operator""_MiB(unsigned long long mebibytes)
+namespace util::detail {
+template<unsigned SHIFT>
```
Comment (hodlinator) [inline src/util/byte_units.h]: nit: I suspect clang-format prefers:
```suggestion
template <unsigned SHIFT>
```

```
$ git grep "template <" src/ | wc -l
807
$ git grep "template<" src/ | wc -l
424
```

**Review by hodlinator — CHANGES_REQUESTED** [review]

File: src/util/byte_units.h (hunk 1 above)
Comment (l0rinc) [reply src/util/byte_units.h]: I had that originally but had some problems, don't remember what. Will try again on next push.
Edit: it's used in `std::overflow_error` which expects a `String`:
```C++
  _LIBCPP_HIDE_FROM_ABI explicit overflow_error(const string& __s) : runtime_error(__s) {}
  _LIBCPP_HIDE_FROM_ABI explicit overflow_error(const char* __s) : runtime_error(__s) {}
```

File: src/util/byte_units.h (hunk 3 above)
Comment (l0rinc) [reply src/util/byte_units.h]: Thanks, will do it on next push

Edit: pushed this nit + rebased

File: src/qt/optionsmodel.h (hunk 2 above)
Comment (l0rinc) [reply src/qt/optionsmodel.h:30]: https://github.com/bitcoin/bitcoin/blob/25884bd89684262b94e09c9904c1560d5ba87d86/src/qt/intro.cpp#L35 indicates this cannot be negative

File: src/util/byte_units.h (hunk 1 above)
Comment (hodlinator) [reply src/util/byte_units.h]: Yes, so why not use `const char*` to avoid the "heap" allocation (usually in constexpr/consteval) in cases where the function succeeds and the exception isn't even thrown?

Comment (l0rinc) [reply src/util/byte_units.h]: That would be true without `constexpr`, as shown in https://godbolt.org/z/TGb57r7zz.
But with `constexpr` it has zero runtime impact since the exception string is completely optimized away  e.g. `8_MiB` and friends are always compile-time constants, see https://godbolt.org/z/7rhfqc39f.
But it would eliminate a string import, so I don't mind, pushed.

File: src/init.cpp [hunk 4]
```diff
@@ -1919,7 +1920,7 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
 
     // On first startup, warn on low block storage space
     if (!do_reindex && !do_reindex_chainstate && chain_active_height <= 1) {
-        uint64_t assumed_chain_bytes{chainparams.AssumedBlockchainSize() * 1024 * 1024 * 1024};
+        uint64_t assumed_chain_bytes{chainparams.AssumedBlockchainSize() * 1_MiB * 1024};
```
Comment (hodlinator) [inline src/init.cpp]: In 112b9914639aab450f0aadc187f6985f56cb7e70: How about putting the GiB commit first and only touch this line once instead of twice?

File: src/qt/optionsmodel.h (hunk 2 above)
Comment (hodlinator) [reply src/qt/optionsmodel.h:30]: Experimented a bit and it seems behavior hasn't changed. Passing test:

```C++
constexpr uint64_t GB_BYTES{1000000000};
static inline int PruneMiBtoGB_old(int64_t mib) { return (mib * 1024 * 1024 + GB_BYTES - 1) / GB_BYTES; }
static inline int PruneMiBtoGB_new(int64_t mib) { return (mib * 1_MiB + GB_BYTES - 1) / GB_BYTES; }
static inline int64_t PruneGBtoMiB_old(int gb) { return gb * GB_BYTES / 1024 / 1024; }
static inline int64_t PruneGBtoMiB_new(int gb) { return gb * GB_BYTES / 1_MiB; }

BOOST_AUTO_TEST_CASE(prune_tests)
{
    BOOST_CHECK_EQUAL(PruneMiBtoGB_old(1500), 2);
    BOOST_CHECK_EQUAL(PruneMiBtoGB_new(1500), 2);
    BOOST_CHECK_EQUAL(PruneMiBtoGB_old(-1500), 1266874889);
    BOOST_CHECK_EQUAL(PruneMiBtoGB_new(-1500), 1266874889);

    BOOST_CHECK_EQUAL(PruneGBtoMiB_old(1), 953);
    BOOST_CHECK_EQUAL(PruneGBtoMiB_new(1), 953);
    BOOST_CHECK_EQUAL(PruneGBtoMiB_old(-1), 17592186043462);
    BOOST_CHECK_EQUAL(PruneGBtoMiB_new(-1), 17592186043462);
}
```

**Review by hodlinator — COMMENTED** [review]
Concept ACK e5b573f18ac2e13508330ada6e92c8155f86a0e5

Makes the code more readible.

Examples:
```diff
-constexpr unsigned int FLTR_FILE_CHUNK_SIZE = 0x100000; // 1 MiB
+constexpr unsigned int FLTR_FILE_CHUNK_SIZE = 1_MiB;

-if (m_total_allocated > 0x1000000) return;
+if (m_total_allocated > 16_MiB) return;

-cuckoo_cache.setup_bytes(megabytes << 20);
+cuckoo_cache.setup_bytes(megabytes * 1_MiB);

-LogDebug(BCLog::COINDB, "Writing final batch of %.2f MiB\n", batch.ApproximateSize() * (1.0 / 1048576.0));
+LogDebug(BCLog::COINDB, "Writing final batch of %.2f MiB\n", batch.ApproximateSize() / double(1_MiB));

-constexpr uint64_t min_disk_space = 52428800; // 50 MiB
+constexpr uint64_t min_disk_space = 50_MiB;
```

File: src/init.cpp (hunk 4 above)
Comment (l0rinc) [reply src/init.cpp]: Good idea, haven't thought of that! Pushed - also simplified the [`gib_string_literal_test`](https://github.com/bitcoin/bitcoin/compare/e5b573f18ac2e13508330ada6e92c8155f86a0e5..792d679a54dcaf359628554fa2681c03d05f1676) to remove some duplication.

File: src/test/util_tests.cpp [hunk 5]
```diff
@@ -1827,10 +1827,69 @@ BOOST_AUTO_TEST_CASE(saturating_left_shift_test)
...
+    BOOST_CHECK_EQUAL((1_GiB + 123) / double(1_GiB), (1_GiB + 123) / 1024.0 / 1024.0 / 1024.0);
+    BOOST_CHECK_EQUAL(2ULL * 1_GiB, 2ULL << 30);
+    BOOST_CHECK_EQUAL(2_GiB, 2048_MiB);
+    BOOST_CHECK_EQUAL(3_GiB / 1_GiB, 3U);
+    BOOST_CHECK_EQUAL(3_GiB, 3U << 30);
+
+    // Overflow handling and specific codebase values
+    constexpr auto max_gib{std::numeric_limits<size_t>::max() >> 30};
+    if constexpr (SIZE_MAX == UINT32_MAX) {
+        BOOST_CHECK_EQUAL(max_gib, 3U);
```
Comment (hodlinator) [inline src/test/util_tests.cpp:1922]: remark:
* Initial thought: "`(2^32) >> 30` should be `2^2 == 4`, not `3`".
* Then I realized `uint32_t`-max is `(2^32)-1` and rounds down.

**Review by hodlinator — APPROVED** [review]
ACK 792d679a54dcaf359628554fa2681c03d05f1676

Concpt. review: https://github.com/bitcoin/bitcoin/pull/34435#pullrequestreview-3724982952

Checked trickier hex literals actually match _MiB/_GiB-versions.

Diffed against previously reviewed push and aside from commit re-ordering, only saw minimal deduplication in gib_string_literal_test.

**Review by hodlinator — APPROVED** [review]
re-ACK 6091d9105d88ca4a43385dcb1936cf9a0b6c6b03

Change since previous review: Adjusts the second commit to merge cleanly after surrounding code changed, mainly #includes.

**Review by hodlinator — APPROVED** [review]
re-ACK 6069bc9de1ff21db304096b6b5fdfb84d8506ad0

Just changes to surrounding diff context and `WARN_FLUSH_COINS_SIZE` being replaced by constant no longer measured in bytes.

**Review by hodlinator — APPROVED** [review]
re-ACK c8e955ca83750eb85873c4823ad1d2128cf9f28b

(Was just rebased to kick CI).

File: src/test/util_tests.cpp [hunk 6]
```diff
@@ -1827,10 +1827,69 @@ BOOST_AUTO_TEST_CASE(saturating_left_shift_test)
...
+        BOOST_CHECK_EXCEPTION(4_GiB, std::overflow_error, HasReason("GiB value too large for size_t byte conversion"));
```
Comment (maflcko) [inline src/test/util_tests.cpp:1923]: would be nice to add:

```cpp
        BOOST_CHECK_EQUAL(4*uint64_t{1_GiB}, uint64_t{4} << 30);
```

An alternative would be to just make GiB return u64, and possibly MiB as well.

There shouldn't be any call-site that *really* needs size_t. size_t seems like  a mental overhead, providing bugs such as https://github.com/bitcoin/bitcoin/pull/34109 and https://github.com/bitcoin/bitcoin/pull/33724

Comment (l0rinc) [reply src/test/util_tests.cpp:1923]: Thanks for the suggestion, 32-bit architectures should also support this widening-first approach, so I [added](https://github.com/bitcoin/bitcoin/compare/c8e955ca83750eb85873c4823ad1d2128cf9f28b..02184dedebe15734171e772853c5d62c03b80294#diff-2bd57d7fbec4bb262834d155c304ebe15d26f73fea87c75ff273df3529a15510R1876) it to the test suite.

I dislike `size_t` too (for the exact reasons you mentioned), and I’d gladly review a PR that addresses that, but it seems more involved than this one.

Comment (maflcko) [reply src/test/util_tests.cpp:1923]: About size_t being annoying, see also https://github.com/bitcoin/bitcoin/pull/34692#issuecomment-3973281952 :sweat_smile:

File: src/test/util_tests.cpp [hunk 7]
```diff
@@ -1828,10 +1828,37 @@ BOOST_AUTO_TEST_CASE(saturating_left_shift_test)
...
+    BOOST_CHECK_EQUAL(4_MiB, 1 << 22);
+    BOOST_CHECK_EQUAL(8_MiB, 1 << 23);
+    BOOST_CHECK_EQUAL(16_MiB, 0x1000000U);
+    BOOST_CHECK_EQUAL(16_MiB, 1 << 24);
+    BOOST_CHECK_EQUAL(32_MiB, 0x2000000U);
+    BOOST_CHECK_EQUAL(32_MiB, 32U << 20);
+    BOOST_CHECK_EQUAL(50_MiB / 1_MiB, 50U);
+    BOOST_CHECK_EQUAL(50_MiB, 52428800U);
+    BOOST_CHECK_EQUAL(128_MiB, 0x8000000U);
+    BOOST_CHECK_EQUAL(550_MiB, 550ULL * 1024 * 1024);
```
Comment (sedited) [inline src/test/util_tests.cpp:1851]: Are these really useful? Feels a bit like testing `BOOST_CHECK_EQUAL(1+1,2)`.

**Review by sedited — COMMENTED** [review]
Concept ACK

I'm not sure about the first commit, seems a bit noisy to do a conversion from 1024 MiB to 1 GiB for a developer who is probably aware of that fact since shortly after using a computer for the first time. For me at least it is also a good reminder that the unit of  `-dbcache` is MiBs.

The second commit seems like a decent readability improvement though.

File: src/test/util_tests.cpp (hunk 7 above)
Comment (l0rinc) [reply src/test/util_tests.cpp:1851]: The reviewers can decide. I covered every change category that I modified in this commit as proof that it doesn't change behavior. I often err on the side of over-testing.

**Comment by l0rinc** [comment]
Rebased after https://github.com/bitcoin/bitcoin/pull/34448/changes#diff-21abb6b14af1e9330a6f0c89a87231035a439248c556ef5e110eb0617b88a1f4R11, ready for review again.

Edit: had some problems with the new IWYU rules

**Comment by hodlinator** [comment]
re-ACK a35f04a76a26da58f4168014518783633115697d

Since last review:
* Added 4 GiB uint64_t test in gib_string_literal_test as suggested in https://github.com/bitcoin/bitcoin/pull/34435#discussion_r2859273735
* Otherwise just #include and diff context updates.

File: src/init.cpp [hunk 8]
```diff
@@ -1831,18 +1832,18 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
...
 
     LogInfo("Cache configuration:");
-    LogInfo("* Using %.1f MiB for block index database", kernel_cache_sizes.block_tree_db * (1.0 / 1024 / 1024));
+    LogInfo("* Using %.1f MiB for block index database", kernel_cache_sizes.block_tree_db / double(1_MiB));
     if (args.GetBoolArg("-txindex", DEFAULT_TXINDEX)) {
-        LogInfo("* Using %.1f MiB for transaction index database", index_cache_sizes.tx_index * (1.0 / 1024 / 1024));
+        LogInfo("* Using %.1f MiB for transaction index database", index_cache_sizes.tx_index / double(1_MiB));
     }
     if (args.GetBoolArg("-txospenderindex", DEFAULT_TXOSPENDERINDEX)) {
         LogInfo("* Using %.1f MiB for transaction output spender index database", index_cache_sizes.txospender_index * (1.0 / 1024 / 1024));
```
Comment (w0xlt) [inline src/init.cpp:1840]: Likely `txospenderindex` was merged between rebases and the rebase resolution kept the old pattern.

```diff
     if (args.GetBoolArg("-txospenderindex", DEFAULT_TXOSPENDERINDEX)) {
-        LogInfo("* Using %.1f MiB for transaction output spender index database", index_cache_sizes.txospender_index * (1.0 / 1024 / 1024));
+        LogInfo("* Using %.1f MiB for transaction output spender index database", index_cache_sizes.txospender_index / double(1_MiB));
     }
```

File: src/node/blockstorage.h
Comment (w0xlt) [inline src/node/blockstorage.h:21]: For IWYU reason (`src/node/blockstorage.h` gets it transitively through `<dbwrapper.h>`).

```diff
 #include <util/expected.h>
+#include <util/byte_units.h>
 #include <util/fs.h>
```

File: src/node/caches.cpp [hunk 9]
```diff
@@ -20,15 +20,15 @@
...
-static constexpr size_t MAX_TXOSPENDER_INDEX_CACHE{1024_MiB};
+static constexpr size_t MAX_TXOSPENDER_INDEX_CACHE{1_GiB};
 //! Maximum dbcache size on 32-bit systems.
-static constexpr size_t MAX_32BIT_DBCACHE{1024_MiB};
+static constexpr size_t MAX_32BIT_DBCACHE{1_GiB};
 //! Larger default dbcache on 64-bit systems with enough RAM.
-static constexpr size_t HIGH_DEFAULT_DBCACHE{1024_MiB};
+static constexpr size_t HIGH_DEFAULT_DBCACHE{1_GiB};
 //! Minimum detected RAM required for HIGH_DEFAULT_DBCACHE.
 static constexpr uint64_t HIGH_DEFAULT_DBCACHE_MIN_TOTAL_RAM{4096ULL << 20};
```
Comment (w0xlt) [inline src/node/caches.cpp:33]: For consistency,

```diff
 //! Minimum detected RAM required for HIGH_DEFAULT_DBCACHE.
-static constexpr uint64_t HIGH_DEFAULT_DBCACHE_MIN_TOTAL_RAM{4096ULL << 20};
+static constexpr uint64_t HIGH_DEFAULT_DBCACHE_MIN_TOTAL_RAM{uint64_t{4} * 1_GiB};
```

**Review by w0xlt — COMMENTED** [review]
ACK a35f04a76a26da58f4168014518783633115697d mod below nits.

File: src/test/util_tests.cpp (hunk 6 above)
Comment (maflcko) [reply src/test/util_tests.cpp:1923]: > I dislike `size_t` too (for the exact reasons you mentioned), and I’d gladly review a PR that addresses that, but it seems more involved than this one.

Why would it be more involved? For me, it is just a trivial patch replacing the type, see https://github.com/bitcoin/bitcoin/pull/35097.

The most involved thing was to play code-golf on the test code, until all compilers accepted it 🫠 

Note, I am not asking to change all `size_t` to `uint64_t` in the whole codebase in one go, but only allow u64 in the return value for this one operator.

**Review by maflcko — COMMENTED** [review]
left a comment

**Comment by l0rinc** [comment]
rfm?

**Comment by maflcko** [comment]
> rfm?

I am not against merging it, but I don't see the point in modifying all the lines twice. Let's recall that the goal here is to check the range inside the operator.

Then, changing  a line to use the operator like `uint32_t bla = 4096_MiB;` (without the C++11 narrowing check) seems fragile, because it silently compiles without warning.

It would be better to   use the C++11 narrowing check when touching. Otherwise this just looks like a style change, rather than a hardening change.

**Comment by l0rinc** [comment]
Thanks for the review, @maflcko, @w0xlt!

Applied @maflcko's braced-initialization suggestion as a new third commit so the narrowing-check hardening can be reviewed independently from the `_MiB` refactor (one site in `chainstatemanager_args.cpp` needed an explicit `size_t(...)` cast to silence a 32-bit narrowing).

Folded @w0xlt's three suggestions (the missed `txospender_index` log-line conversion, the `<util/byte_units.h>` IWYU include in `node/blockstorage.h`, and `uint64_t{4} * 1_GiB` for `HIGH_DEFAULT_DBCACHE_MIN_TOTAL_RAM`) into the second commit.

```bash
git range-diff 459f813bd4890e8842e8840a42caf943481cc4be..a35f04a76a26da58f4168014518783633115697d c5ef940129f247388e809cb99347519255487edf..c6bb8a132a3def8b6d11a0382152602806d1e241
```

File: src/bench/chacha20.cpp [hunk 10]
```diff
@@ -15,7 +16,7 @@
 /* Number of bytes to process per iteration */
 static const uint64_t BUFFER_SIZE_TINY  = 64;
 static const uint64_t BUFFER_SIZE_SMALL = 256;
-static const uint64_t BUFFER_SIZE_LARGE = 1024*1024;
+static const uint64_t BUFFER_SIZE_LARGE = 1_MiB;
```
Comment (maflcko) [inline src/bench/chacha20.cpp]: i think it is fine to squash the third commit into the earlier ones. Otherwise, there is extra work for reviewers, because they have to review twice and also  go back and forth and check that the second time, all places of the first change were indeed touched again.

If you want to split up commits, you can probably split up the `src/util/byte_units.h` changes (+unit tests) into the first commit, but no strong opinion, up to you.

Comment (l0rinc) [reply src/bench/chacha20.cpp]: Sure, folded the changes back to the original commits

File: src/qt/optionsmodel.h (hunk 2 above)
Comment (janb84) [inline src/qt/optionsmodel.h:30]:
```suggestion
static inline int PruneMiBtoGB(int64_t mib) { 
   Assume(mib >=0)
   return (mib * 1_MiB + GB_BYTES - 1) / GB_BYTES; 
}
```

NIT (very optional): You could make it a bit more explicit why the arithmetic isn't problematic by adding the `Assume` precondition assertion.

**Review by janb84 — COMMENTED** [review]
ACK af0ee28eb623a54e7ff8602864f61cab1f23848d

Using GiB next to MiB seems like a logical next step, data usages are increasing.
And consistently using MiB/Gib makes the code easier to reason about. (1_MiB in stead of (1024*1024))

Small very optional NIT, for the rest the change looks good to me.

File: src/node/caches.cpp [hunk 11]
```diff
@@ -20,17 +20,17 @@
...
+static constexpr size_t MAX_TXOSPENDER_INDEX_CACHE{1_GiB};
 //! Maximum dbcache size on 32-bit systems.
-static constexpr size_t MAX_32BIT_DBCACHE{1024_MiB};
+static constexpr size_t MAX_32BIT_DBCACHE{1_GiB};
 //! Larger default dbcache on 64-bit systems with enough RAM.
-static constexpr size_t HIGH_DEFAULT_DBCACHE{1024_MiB};
+static constexpr size_t HIGH_DEFAULT_DBCACHE{1_GiB};
 //! Minimum detected RAM required for HIGH_DEFAULT_DBCACHE.
-static constexpr uint64_t HIGH_DEFAULT_DBCACHE_MIN_TOTAL_RAM{4096ULL << 20};
+static constexpr uint64_t HIGH_DEFAULT_DBCACHE_MIN_TOTAL_RAM{uint64_t{4} * 1_GiB};
```
Comment (maflcko) [inline src/node/caches.cpp:33]: nit in b3edd30aa24af3a5f63cbcebb7013e60a6d7b8a4: Looks a bit like a shortcoming that should be fixed up after merge, or this should wait for https://github.com/bitcoin/bitcoin/pull/35097

File: src/node/blockstorage.h [hunk 12]
```diff
@@ -18,6 +18,7 @@
 #include <streams.h>
 #include <sync.h>
 #include <uint256.h>
+#include <util/byte_units.h> // IWYU pragma: keep
```
Comment (maflcko) [inline src/node/blockstorage.h:21]: nit in af0ee28eb623a54e7ff8602864f61cab1f23848d: Obviously nothing that can be done here, but after merge, it could make sense to open a tracking issue about this. Maybe iwyu is confusing itself with includes that are already present indirectly?

File: src/node/chainstatemanager_args.cpp [hunk 13]
```diff
@@ -64,7 +65,7 @@ util::Result<void> ApplyArgsManOptions(const ArgsManager& args, ChainstateManage
         //    script execution cache create the minimum possible cache (2
         //    elements). Therefore, we can use 0 as a floor here.
         // 2. Multiply first, divide after to avoid integer truncation.
-        size_t clamped_size_each = std::max<int64_t>(*max_size, 0) * (1 << 20) / 2;
+        size_t clamped_size_each{size_t(std::max<int64_t>(*max_size, 0) * 1_MiB / 2)};
```
Comment (maflcko) [inline src/node/chainstatemanager_args.cpp:68]: nit in af0ee28eb623a54e7ff8602864f61cab1f23848d: I know that maxsigcachesize is a debug-only option for experts, but since you are adding a cast here, and given that can overflow, see also commit ab14d1d6a4a8ef5fe5013150e6c5ebcb5f5e4ea9, it seems better to avoid the overflow.

I think it would be fine to call `GetArg<uint32_t>("-maxsigcachesize")` here, which allows to drop the `max` call, and the result should fit into u64, which can then be passed to `setup_bytes`, which will then apply the `min` call.

File: src/randomenv.cpp [hunk 14]
```diff
@@ -113,7 +114,7 @@ void AddFile(CSHA512& hasher, const char *path)
             if (n > 0) hasher.Write(fbuf, n);
             total += n;
             /* not bothering with EINTR handling. */
-        } while (n == sizeof(fbuf) && total < 1048576); // Read only the first 1 Mbyte
+        } while (n == sizeof(fbuf) && total < 1_MiB); // Read only the first 1 Mbyte
```
Comment (maflcko) [inline src/randomenv.cpp:117]: nit from the llm in af0ee28eb623a54e7ff8602864f61cab1f23848d: `// Read only the first 1 Mbyte -> // Read only the first 1 MiB [The code uses 1_MiB (mebibytes), but the comment says “Mbyte” (megabytes), which is a different unit and could mislead understanding of the cutoff size.]`

File: src/test/blockfilter_index_tests.cpp [hunk 15]
```diff
@@ -277,14 +278,14 @@ BOOST_FIXTURE_TEST_CASE(blockfilter_index_init_destroy, BasicTestingSetup)
     filter_index = GetBlockFilterIndex(BlockFilterType::BASIC);
     BOOST_CHECK(filter_index == nullptr);
 
-    BOOST_CHECK(InitBlockFilterIndex([&]{ return interfaces::MakeChain(m_node); }, BlockFilterType::BASIC, 1 << 20, true, false));
+    BOOST_CHECK(InitBlockFilterIndex([&]{ return interfaces::MakeChain(m_node); }, BlockFilterType::BASIC, 1_MiB, true, false));
```
Comment (maflcko) [inline src/test/blockfilter_index_tests.cpp:281]: llm nit in the second commit: I know this is unrelated, but it would be nice to use named args for those bools, whose meaning and order is otherwise unclear. Either inside the second commit, or a new commit.

```
Possible places where named args for integral literals may be used (e.g. func(x, /*named_arg=*/0) in C++, and func(x, named_arg=0) in Python):

    BlockFilterIndex filter_index(interfaces::MakeChain(m_node), BlockFilterType::BASIC, 1_MiB, true) in src/test/blockfilter_index_tests.cpp
    InitBlockFilterIndex([&]{ return interfaces::MakeChain(m_node); }, BlockFilterType::BASIC, 1_MiB, true, false) in src/test/blockfilter_index_tests.cpp
    InitBlockFilterIndex([&]{ return interfaces::MakeChain(m_node); }, BlockFilterType::BASIC, 1_MiB, true, false) in src/test/blockfilter_index_tests.cpp
    CoinStatsIndex coin_stats_index{interfaces::MakeChain(m_node), 1_MiB, true} in src/test/coinstatsindex_tests.cpp
    TxIndex txindex(interfaces::MakeChain(m_node), 1_MiB, true) in src/test/txindex_tests.cpp

**Review by maflcko — COMMENTED** [review]
lgtm, left some comments/nits

review ACK af0ee28eb623a54e7ff8602864f61cab1f23848d 🖍

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: review ACK af0ee28eb623a54e7ff8602864f61cab1f23848d 🖍
7kLrlcqOwueHKhUdFEup7YgZ/HgwqO9GMJqW5/5/Pio6L/rV0Eb9p6jbyR8N0+9w72gj8yKIy/R/WHtPYpl+Dg==
```

</details>

File: src/node/blockstorage.h (hunk 12 above)
Comment (fanquake) [reply src/node/blockstorage.h:21]: The one thing that should be done, is explain why this is needed, in an inline comment.

File: src/node/caches.cpp (hunk 11 above)
Comment (l0rinc) [reply src/node/caches.cpp:33]: This was specifically [requested](https://github.com/bitcoin/bitcoin/pull/34435#discussion_r3068875148) - your follow-up PR will allow us to make this less hacky.

File: src/node/blockstorage.h (hunk 12 above)
Comment (l0rinc) [reply src/node/blockstorage.h:21]: I'm not sure why IWYU is misfiring here, but the file has [obvious `_MiB`](https://github.com/bitcoin/bitcoin/pull/34435/changes#diff-ed3f90693a242b38b9719af171de8f55183576957676dfa358945bea22276bd5R120-R124) usages. Fixing IWYU is outside the scope of this PR - without this line the CI was failing.

File: src/node/chainstatemanager_args.cpp (hunk 13 above)
Comment (l0rinc) [reply src/node/chainstatemanager_args.cpp:68]: Do you mind if we do that in a follow-up instead?

File: src/randomenv.cpp (hunk 14 above)
Comment (l0rinc) [reply src/randomenv.cpp:117]: I thought about removing the comment completely now - but decided to keep it, if I need to touch again I'll just remove it

File: src/test/blockfilter_index_tests.cpp (hunk 15 above)
Comment (l0rinc) [reply src/test/blockfilter_index_tests.cpp:281]: Agree, I will either do it in a follow-up or here, if I need to touch the code again.

File: src/qt/optionsmodel.h (hunk 2 above)
Comment (l0rinc) [reply src/qt/optionsmodel.h:30]: Thanks, if I need to touch again I'll consider it

File: src/node/blockstorage.h (hunk 12 above)
Comment (fanquake) [reply src/node/blockstorage.h:21]: There's no need to fix IWYU, just to document the workaround.

File: src/node/chainstatemanager_args.cpp (hunk 13 above)
Comment (maflcko) [reply src/node/chainstatemanager_args.cpp:68]: Well, I am not sure. I just have a really hard time reviewing this line in the diff. To clarify: The MiB operator may return u32, or u64. However, integer promotion depends on the width of the type.

On master `(1<<20)` is int (i32), multiplied with i64, is i64, which may overflow signed (UB/crash)

With the changes here, `MiB` is either u32, or u64. In one case the multiply will yield u64, in the other i64. So for one platform, this is a UB/crash, for the other platform, it is a silent and defined unsigned overflow.

I guess it is just me, and the other reviewers were fine with this, but I find it tedious and pointless to review every line of code twice for every architecture. With https://github.com/bitcoin/bitcoin/pull/35097, there would only be one consistent behavior and only one review needed.

File: src/node/blockstorage.h (hunk 12 above)
Comment (l0rinc) [reply src/node/blockstorage.h:21]: I have opened a fix in include-what-you-use/include-what-you-use#2014.

The issue was that IWYU could miss that a user-defined literal requires the header declaring its literal operator, and would then incorrectly suggest removing it.

File: src/node/chainstatemanager_args.cpp (hunk 13 above)
Comment (hodlinator) [reply src/node/chainstatemanager_args.cpp:68]: I agree that the current change is awkward as the multiplication leads to different signedness on 32 and 64 bit due to integer promotion rules for same/different sized integers. (Whereas on master the behavior has a higher chance of being consistent). How about moving the ending parens of the cast?
```suggestion
        size_t clamped_size_each{size_t(std::max<int64_t>(*max_size, 0)) * 1_MiB / 2};
```
That way we multiply `size_t * size_t`, which is either `u32 * u32` or `u64 * u64`.

There's still potential for unsigned overflow, but at least it's not UB. (Would almost be nice to have `SaturatingMultiply()`).

Edit: I guess even with my change on 32-bit, `max_size` could be larger than max u32 and therefore still lead to UB when casting to `size_t`. :\

Comment (l0rinc) [reply src/node/chainstatemanager_args.cpp:68]: Should I just drop this line and we'll fix it in another PR?

File: src/node/caches.cpp (hunk 11 above)
Comment (hodlinator) [reply src/node/caches.cpp:33]: Things like this line makes me prefer that #35097 be merged first even if it was created as a reaction to this one. Improving the underlying primitive before applying it in more places makes more sense. But we can fix it up later.

File: src/node/chainstatemanager_args.cpp (hunk 13 above)
Comment (hodlinator) [reply src/node/chainstatemanager_args.cpp:68]: It would feel good with a more complete fix, clamping the input somehow to avoid overflow. Or dropping it from the PR. But if you don't re-touch the PR for other reasons, it can be done in a follow-up IMO.

**Review by hodlinator — APPROVED** [review]
re-ACK af0ee28eb623a54e7ff8602864f61cab1f23848d

Good to see the anti-narrowing braces.

**Comment by achow101** [comment]
ACK af0ee28eb623a54e7ff8602864f61cab1f23848d

File: src/node/blockstorage.h (hunk 12 above)
Comment (hebasto) [reply src/node/blockstorage.h:21]: > I have opened a fix in [include-what-you-use/include-what-you-use#2014](https://github.com/include-what-you-use/include-what-you-use/pull/2014).

Thank you!

Comment (hebasto) [reply src/node/blockstorage.h:21]: Leaving a note for myself. The IWYU master branch @ [`ae18593bd2f3d4b16d8e9921aa95af0afb368490`](https://github.com/include-what-you-use/include-what-you-use/commit/ae18593bd2f3d4b16d8e9921aa95af0afb368490) requires the following change:
```diff
--- a/src/index/blockfilterindex.cpp
+++ b/src/index/blockfilterindex.cpp
@@ -19,6 +19,7 @@
 #include <sync.h>
 #include <uint256.h>
 #include <util/check.h>
+#include <util/byte_units.h>
 #include <util/fs.h>
 #include <util/hasher.h>
 #include <util/log.h>
```

Comment (l0rinc) [reply src/node/blockstorage.h:21]: The reproducer I used adjusted more than that: https://github.com/l0rinc/bitcoin/pull/141/changes/63325eacd6aef3dc76736ff0b8526bf680b101af
